include/envoy/stream_info/filter_state.h

```diff
@@ -50,6 +57,23 @@ class FilterState {
     return *result;
   }
 
+  /**
+   * @param data_name the name of the data being looked up.
+   * @return a non-const reference to the stored data if and only if the
+   * underlying data is mutable.
+   * An exception will be thrown if the data does not exist or if it is
+   * immutable. This function will fail if the data stored under
+   * |data_name| cannot be dynamically cast to the type specified.
+   */
+  template <typename T> T& getData(absl::string_view data_name) {
+    T* result = dynamic_cast<T*>(getDataGeneric(data_name));
+    if (!result) {
+      throw EnvoyException(
```

Comment:
I don't think we should throw exception here given this is mostly used on the data path per new added style guide. @htuch thoughts?

Reply:
I think it's fine to throw an exception here. IMO this is essentially a programming error and should cause the server to crash. The real question is whether we expect this function to be called when the caller does not know the type, and I don't think this will ever be true.

Reply:
CC @curiouserrandy FYI on exception philosophy (also see recently merged https://github.com/envoyproxy/envoy/pull/4793).

Reply:
@mattklein123 why not RELEASE_ASSERT instead of throwing an exception if that is the intended behavior?

Reply:
why do I always get caught with this re-evaluation? Didn't you guys argue and decide about this when @curiouserrandy merged his PR ? All I am doing is *literally* copy pasting the existing getData without const. :(

Reply:
Because programming wouldn't be programming without pointless bikeshedding? ;)

Honestly I don't care either way.

Reply:
Yes, I agree that any changes to the behavior here should be a separate PR, too many separate concerns otherwise.